From 71ee3f120c0500a2723ca162a1e19e8d65975f90 Mon Sep 17 00:00:00 2001
From: Varun Sharma <vsharma@habana.ai>
Date: Tue, 15 Feb 2022 14:46:32 +0200
Subject: [PATCH] [SW-67835] greco: set arc's to privileged in ppboot start

This patch moves make arc's privilege setting to early ppboot init,
instead of doing it in preboot. By default arc's come as
non privilege, and we set them as privilege, so they can access
privilege registers also.

Change-Id: I77c8e4aa82bbc56a2879b1603071d7b18d05cd25
Signed-off-by: Varun Sharma <vsharma@habana.ai>
---

diff --git a/arch/greco/arc/security.c b/arch/greco/arc/security.c
index 4255fa3..8b1fbc3 100644
--- a/arch/greco/arc/security.c
+++ b/arch/greco/arc/security.c
@@ -12,32 +12,7 @@
 #include <sha256.h>
 #include <greco/greco_security.h>
 
-#include <psoc_arc0_mstr_if_core_hbw_regs.h>
-#include <psoc_arc0_mstr_if_core_lbw_regs.h>
-#include <psoc_arc1_mstr_if_core_hbw_regs.h>
-#include <psoc_arc1_mstr_if_core_lbw_regs.h>
-
-#define EN_ARC_SEC_PRIV				0x3
 #define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
-
-void preboot_secure_arc(void)
-{
-	/* enable security and privilege mode for arc0 and arc1 */
-	regmap_writel(EN_ARC_SEC_PRIV,
-		mmPSOC_ARC0_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-	regmap_writel(EN_ARC_SEC_PRIV,
-		 mmPSOC_ARC0_MSTR_IF_CORE_LBW_AXPROT_OVRD);
-	regmap_writel(EN_ARC_SEC_PRIV,
-		 mmPSOC_ARC1_MSTR_IF_CORE_HBW_AXPROT_OVRD);
-	regmap_writel(EN_ARC_SEC_PRIV,
-		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
-}
-
-void preboot_security_init(void)
-{
-	/* TODO add check for security strap ON */
-	preboot_secure_arc();
-}
 
 u32 sp_sram_readl(u32 offset)
 {
diff --git a/arch/greco/arc_elf32/common.c b/arch/greco/arc_elf32/common.c
index e1e6e62..41b14d8 100644
--- a/arch/greco/arc_elf32/common.c
+++ b/arch/greco/arc_elf32/common.c
@@ -27,6 +27,11 @@
 #include <dcon3_hbw_rtr_if0_rtr_h3_regs.h>
 #include <dcon3_hbw_rtr_if1_rtr_h3_regs.h>
 
+#include <psoc_arc0_mstr_if_core_hbw_regs.h>
+#include <psoc_arc0_mstr_if_core_lbw_regs.h>
+#include <psoc_arc1_mstr_if_core_hbw_regs.h>
+#include <psoc_arc1_mstr_if_core_lbw_regs.h>
+
 #include <psoc_arc0_cfg_regs.h>
 #include <psoc_arc0_cfg_masks.h>
 #include <psoc_arc1_cfg_regs.h>
@@ -315,3 +320,18 @@
 	*secondary_iccm_addr = (u32) &secondary_entry;
 	run_arc(0);
 }
+
+#define HL_SEC_EN_ARC_SEC_PRIV         0x3
+
+void ppboot_secure_arc(void)
+{
+	/* enable security and privilege mode for arc0 and arc1 */
+	regmap_writel(HL_SEC_EN_ARC_SEC_PRIV,
+		mmPSOC_ARC0_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(HL_SEC_EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC0_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+	regmap_writel(HL_SEC_EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_HBW_AXPROT_OVRD);
+	regmap_writel(HL_SEC_EN_ARC_SEC_PRIV,
+		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
+}
